Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/TextBoxLikePropertyCtrl.hxx

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <AzCore/std/limits.h>
+#include <AzCore/std/optional.h>
+
+namespace AzToolsFramework
+{
+    struct DeferredTextboxLikeEdit_default_traits
+    {
+        template<class C>
+        using value_type = decltype(AZStd::declval<const C&>().value());
+
+        template<class C>
+        static auto get(const C* obj)
+        {
+            return obj->value();
+        }
+
+        template<class C, class T>
+        static void set(C* obj, const T& val)
+        {
+            obj->setValue(val);
+        }
+
+        template<class C>
+        static bool isUserEditing(const C* obj)
+        {
+            return obj->hasFocus();
+        }
+
+        template<class C>
+        static void emitEditingFinished(C* obj)
+        {
+            emit obj->editingFinished();
+        }
+    };
+
+    /*
+    A small helper for deferring `setValue' events while a user is editing a textbox-like property.
+    In these cases, any external `setValue' events shouldn't overwrite what the user enters.
+    */
+    template<class Derived, class T, class Traits = DeferredTextboxLikeEdit_default_traits>
+    class DeferredTextboxLikeEdit
+    {
+    public:
+        void setValueFromSystem(const T&);
+
+    protected:
+        template<class... Args>
+        void onTextBoxLikeEditingFinished(Args&&... args);
+
+        // Call to check if you should really set the new value
+        // This might record that value for later use
+        bool shouldReallySetValue(const T& prevValue, const T& newValue, bool isUserEditing);
+
+        // Call when the user is done editing the textbox-like property
+        // If this returns a value, you should set the property widget to have that value
+        AZStd::optional<T> getDeferredValuedToSet(const T& value, bool isUserEditing);
+        AZStd::optional<T> getDeferredValuedToSet();
+
+    private:
+        const Derived* GetDerived() const { return static_cast<const Derived*>(this); }
+              Derived* GetDerived()       { return static_cast<      Derived*>(this); }
```

Comment:
nit: I'd avoid this type of formatting, does not hold up well during maintenance (can just run clang-format on this new file which would be great! 🙂)